libcnb/src/layer_env.rs

```diff
@@ -875,6 +875,40 @@ mod test {
     }
 }
 
+pub struct LayerEnvBuilder {
+    layer_env: LayerEnv,
+}
+
+impl LayerEnvBuilder {
```

Comment:
This is one of the cases we spoke about where I'm not sure what value a unit test provides. The code is trivial and has no branches at all. Testing would boil down to writing down what has been written already. 

Since the compiler enforces that the structure is correct at all times, the only "logic" we can test is that the parameters to `with` are passed though: 

```rust
self.layer_env.insert(target, modification_behavior, name, value);
```

If `LayerEnvBuilder` grows out of this trivial state, we can still add unit tests at that point. I'm open for discussions of course and happy to be convinced otherwise. :)

Reply:
I agree that we don't need to test anything that's enforced by the type system. However as you say, the `.insert()` call isn't enforced by the type system. 

> Testing would boil down to writing down what has been written already.

So I really do get this argument, and I'm still trying to work out how it should be applied in practice myself. However I feel like it could be over-applied to many situations, if additional considerations were not factored in. Ultimately all tests are just "writing down what has been written already".

Also, in most cases the tests represent what has been written in a different form - a form that may be easier to parse for correctness, or at least having two forms that can be inspected in the hope that a bug would be apparent in at least one of the forms.

There are also fringe benefits of tests (particularly doctests) other than correctness - eg: demonstrating API use / showing example inputs and outputs, that are not as easily gleaned from reading only the implementation.

Reply:
At the very least we need rustdocs here, so perhaps the answer is rustdocs with a single doctest, then we both show how to use `LayerEnvBuilder` and have a test for `.insert()`.

Reply:
I'm not arguing about adding some RustDoc for it and we might just get a test in via that. :)

When I say "writing down what has been written already", I mean it quite literally. Sure, code ultimately represents its results when executed, but that isn't the point. For example, when code has different branches, we want to test that it takes the branches for a given set of inputs, especially when they're nested. In this case here, `with` is awfully close to be just an alias for another function (which has tests since it actually does things, BTW).

I looked around to find something written by someone who can explain my view more eloquently and I found https://softwareengineering.stackexchange.com/a/429856 which I agree with. I'm not saying that this is the only correct view on the subject, just something that confirms my bias and might explain my point better than I can. ;-)

Reply:
I agree it's very close to an alias. I guess my main concern was that for end users of libcnb, the `LayerEnvBuilder` is the interface they will interact with, not `LayerEnv`, and yet we have all the docs and tests on the latter, not the former - which makes it feel like we're testing/documenting implementation details not the actual more commonly used API.

Reply:
I opened #165 which allows the easy usage of `LayerEnv` with `LayerResult` without the need of a builder. If #165 gets merged, I'll rebase this PR and remove `LayerEnvBuilder` entirely, resolving the issues in this discussion.

Reply:
This is now resolved via 642a39b